packages/core/src/PartialExecutionUtils/__tests__/findStartNodes.test.ts

```diff
@@ -0,0 +1,376 @@
+// NOTE: Diagrams in this file have been created with https://asciiflow.com/#/
+// If you update the tests, please update the diagrams as well.
+// If you add a test, please create a new diagram.
+//
+// Map
+// 0  means the output has no run data
+// 1  means the output has run data
+// ►► denotes the node that the user wants to execute to
+// XX denotes that the node is disabled
+// PD denotes that the node has pinned data
+
+import { type IPinData, type IRunData } from 'n8n-workflow';
+import { createNodeData, toITaskData } from './helpers';
+import { findStartNodes, isDirty } from '../findStartNodes';
+import { DirectedGraph } from '../DirectedGraph';
+
+describe('isDirty', () => {
+	test("if the node has pinned data it's not dirty", () => {
+		const node = createNodeData({ name: 'Basic Node' });
+
+		const pinData: IPinData = {
+			[node.name]: [{ json: { value: 1 } }],
+		};
+
+		expect(isDirty(node, undefined, pinData)).toBe(false);
+	});
+
+	test("if the node has run data it's not dirty", () => {
+		const node = createNodeData({ name: 'Basic Node' });
+
+		const runData: IRunData = {
+			[node.name]: [toITaskData([{ data: { value: 1 } }])],
+		};
+
+		expect(isDirty(node, runData)).toBe(false);
+	});
+});
+
+describe('findStartNodes', () => {
+	//   ►►
+	//  ┌───────┐
+	//  │trigger│
+	//  └───────┘
+	test('finds the start node if there is only a trigger', () => {
+		const node = createNodeData({ name: 'Basic Node' });
+		const graph = new DirectedGraph().addNode(node);
+
+		const startNodes = findStartNodes(graph, node, node);
+
+		expect(startNodes).toHaveLength(1);
+		expect(startNodes[0]).toEqual(node);
+	});
+
+	//                 ►►
+	//  ┌───────┐     ┌───────────┐
+	//  │trigger├────►│destination│
+	//  └───────┘     └───────────┘
+	test('finds the start node in a simple graph', () => {
+		const trigger = createNodeData({ name: 'trigger' });
+		const destination = createNodeData({ name: 'destination' });
+		const graph = new DirectedGraph()
+			.addNodes(trigger, destination)
+			.addConnection({ from: trigger, to: destination });
+
+		// if the trigger has no run data
+		{
```

Comment:
Oh nice, I haven't seen blocks being used like this before, but this actualy makes this much clearer 👍 Tho I guess they could also be their own test cases, as they test two separate cases 🤔 This is fine tho

Reply:
Yeah, I'm torn as well. 
1. duplicating the setup is a bit ick
2. extracting the setup into a function makes the tests less readable
3. extracting the setup into a `beforeEach` means wrapping everything into a `describe` and seems overkill
4. using `test.each` seems like an OK solution, but for 2 cases it's also a bit overkill, I'll go for this if I have 3 or more cases
5. using blocks

I think I'm sticking with the blocks for now.